Block Library: Add home link block

gwwar opened the pull request:
This adds a new dynamic block that renders a link, backed by `home_url()`. Fixes https://github.com/WordPress/gutenberg/issues/20542

We're keeping this block simple to start. There's no url to edit, and subnesting of navigation links is not implemented.

We're also deferring linking to Reader Options (to modify home page settings) in the block settings panel for a follow up PR. We'll need a generalized approach here to allow folks to override these links with a new filter or slot/fill pattern.

There's a decent amount of copy pasted code needed for color/font support from the navigation parent. I do think this can and should be reduced.

For now, I opted to choose a similar implementation as navigation-link for now, while we consider options for what Navigation Block markup might look like. (See https://github.com/WordPress/gutenberg/issues/29036 )

https://user-images.githubusercontent.com/1270189/117212717-88be7f80-adaf-11eb-8f9a-9c9712321ed8.mp4

### Testing Instructions

- In a site or post editor, add a navigation block
- Clicking on the inserter provides a Home Link block option
- Verify that we cannot add a home link block outside of a navigation block parent
- Try to set some custom font sizes/colors on the navigation block parent. Home link styling should appropriately be affected by these changes.

~~Current work here is a very early WIP, but before I get too deep I have a few questions. We currently favor implementing this as a new block, rather than a navigation link variation:~~

- ~~Should we support this only in Navigation? Or are there use cases for adding this outside of the Navigation block, similar to PageList?~~ A: Only in navigation to start
- ~~Should we support child nesting of NavigationLinks inside of a home block? This will likely duplicate logic and complicate the block.~~ A: Keep it shallow
- ~~Naming: is `core/home` okay, or do we prefer something else?~~ A: `core/home-link`
- ~~The user shouldn't have to manually update the url, as such do we still need to display a link control UI? The only piece they may want to toggle is if the link opens in a new tab.~~ A: No need for link UI

Folks are welcome to jump directly onto this branch if you'd like to collaborate. Just leave a comment or DM me so I'll know to a give heads up before rebasing.

gwwar committed:
Add home link block

gwwar commented:
cc @georgeh @vcanales @jasmussen @mtias

carolinan commented:
> * Should we support this only in Navigation? Or are there use cases for adding this outside of the Navigation block, similar to PageList?

Looking through wpdirectory [search results](https://wpdirectory.net/search/01F3EY3450KQP02T1NCAND4T1K)
It is used for:

- Linking site title and logo (irrelevant)
- Search form action (irrelevant)
- Navigation menu / "nav bars"
- Breadcrumbs

And in one theme I saw it used as a "back to home" link on a  404 page

Do we think that there will be a separate breadcrumb block, or perhaps the navigation block serves as breadcrumbs some time in the future?

> * The user shouldn't have to manually update the url, as such do we still need to display a link control UI? The only piece they may want to toggle is if the link opens in a new tab.

Not being able to edit the url sounds reasonable to me.

jasmussen commented:
This is so cool, thanks for working on this! As a quick test, even in this WIP state, looks good:

<img width="692" alt="Screenshot 2021-04-19 at 08 30 42" src="https://user-images.githubusercontent.com/1204802/115191487-c651b600-a0e9-11eb-9813-6ba8c9b2b459.png">

For when the link is inside navigation, we'll likely need the `wp-block-navigation-link` class even on this block, so it inherits existing nav styles:

<img width="520" alt="Screenshot 2021-04-19 at 08 31 15" src="https://user-images.githubusercontent.com/1204802/115191546-dc5f7680-a0e9-11eb-903b-be14948ac053.png">

To your questions:

> Should we support this only in Navigation? Or are there use cases for adding this outside of the Navigation block, similar to PageList?

Carolina shows some excellent use cases. But I think for this block it's probably best to start with it as Navigation-exclusive only. It should be easy to graduate it to work beyond that one, but it's hard to go the other way.

> Should we support child nesting of NavigationLinks inside of a home block? This will likely duplicate logic and complicate the block.

The question surfaces an added complexity. I recall seeing, somewhere in the github archives, a mockup by @jameskoster which lets you select which page is the home page, and which is the blog page, either in the top centered dropdown or the left admin navigation rail. All that is to say that "home" and "blog", in those situations, are _properties of the page_. 

Here, it's a separate block. Which while definitely useful, it immediately makes me think of the page list block, which ideally can also support the "home" classnames.

In that vein, and purely out of speculation: could the "home" aspect be a variation of a menu item, or a property that is set elsewhere?

> The user shouldn't have to manually update the url, as such do we still need to display a link control UI? The only piece they may want to toggle is if the link opens in a new tab.

Fine to disable!

gziolo commented on packages/block-library/src/home/edit.js:
```diff
@@ -0,0 +1,54 @@
+	const { label } = attributes;
+
+	return (
+		<li { ...blockProps }>
```
> Naming: is core/home okay, or do we prefer something else?

At the moment the implementation isn't very general (because of `li` wrapper), so `core/home` doesn't seem to be a great fit.

gwwar replied:
I didn't show it in this branch, but we could update the behavior such that the li wrapper is only added when the parent is a navigation block. This is similar to what the PageList does

gziolo replied:
Interesting idea to offer an automatic wrapper element.

gziolo commented:
> Should we support child nesting of NavigationLinks inside of a home block? This will likely duplicate logic and complicate the block.

We also have Site Title and Site Logo blocks that both link to the homepage. Would the make sense to be added to the Navigation block. I saw a demo from @mtias where he added the Search Form block to the Navigation block so I would assume that Site Title and Site Logo could be used as well. I'm mostly sharing because there is so many possibilities to take into account 😅

gwwar commented:
From initial feedback here, I suspect we don't have a good enough use case for a brand new block here. Logic would initially be heavily tied to matching navigation links, which might end up being pretty confusing.

I've spun up https://github.com/WordPress/gutenberg/pull/30975 as an alternative using a navigation link variation. Let me know what folks think, especially if y'all feel strongly if this should be a new block.

> Do we think that there will be a separate breadcrumb block

@carolinan thanks for the use case examples. I personally think a breadcrumb block would be pretty handy for folks (and I suspect it'd likely would be implemented separately from the navigation block). I'm not sure if there are current plans for adding a breadcrumb as a new core block.

> We also have Site Title and Site Logo blocks that both link to the homepage.

Good spot @gziolo. 🤔 Would it make sense then that a home link is some type of style variation of site title?

> For when the link is inside navigation, we'll likely need the wp-block-navigation-link class even on this block, so it inherits existing nav styles:

Great feedback @jasmussen

mtias commented:
> Should we support this only in Navigation? Or are there use cases for adding this outside of the Navigation block, similar to PageList?

Just in navigation is fine. The breadcrumb case is a nice one, but I think a specific block for that would be preferred.

> Should we support child nesting of NavigationLinks inside of a home block?

No, I'd keep it simple and shallow.

> Naming: is core/home okay, or do we prefer something else?

`core/home-link`

> The user shouldn't have to manually update the url, as such do we still need to display a link control UI? The only piece they may want to toggle is if the link opens in a new tab.

Correct. No UI needed for editing the url. At most we might want to deep link to permalink settings and such, or to the home template if it's a block theme.

jameskoster commented:
> The question surfaces an added complexity. I recall seeing, somewhere in the github archives, a mockup by @jameskoster which lets you select which page is the home page, and which is the blog page, either in the top centered dropdown or the left admin navigation rail.

https://github.com/WordPress/gutenberg/issues/28341#issuecomment-801911429

gwwar commented:
Leaving a reminder note to myself to update the PageList -> NavigationLinks conversion to convert to a HomeLink block when we detect this case. (It's probably be fine to do as a follow up PR too).

gwwar committed:
rename home block to home-link

gwwar committed:
add support for title rel and target

gwwar commented:
I was digging into the Navigation Block markup a bit and two things stick out:

### We have invalid markup for social links, search, and spacer

Each of these need to be wrapped by an li tag. Folks can use https://validator.w3.org/#validate_by_input to verify this.

I think one way to tidy this up would be to let the Navigation block itself render the top li for each child.

Edit: looks like there's https://github.com/WordPress/gutenberg/issues/29036 for this

### There's a lot of duplicated classes applied to navigation link (and likely needed for home link).

For things like font/color selections on the nav block we add a number classes like `has-text-color` to both the navigation block and links. There is actually a decent amount of code that's copied for the dynamic render_callback, in addition to a lot of CSS that also tries to inherit these styles

I can simply copy paste this for the home link, but I wonder if we can reduce this.

<img width="989" alt="Screen Shot 2021-04-22 at 8 44 39 AM" src="https://user-images.githubusercontent.com/1270189/115744050-0feb0c80-a347-11eb-8def-46abf2ed2efc.png">

gwwar committed:
Add color/font support

jasmussen commented:
> I can simply copy paste this for the home link, but I wonder if we can reduce this.

I would be happy to collaborate on this if you like, or to just volunteer to fix any CSS inheritance of the nav block that needs to be done if something breaks in a simplification.

gwwar commented:
> I would be happy to collaborate on this if you like, or to just volunteer to fix any CSS inheritance of the nav block that needs to be done if something breaks in a simplification.

Your commits are always welcome @jasmussen ✨

gwwar commented:
This one is ready for review

jasmussen committed:
A little polish.

jasmussen commented:
This is a nice, simple block. Works as intended:

![home link](https://user-images.githubusercontent.com/1204802/116987045-da430d80-acce-11eb-9cb9-affb938dbb1c.gif)

I pushed a little polish to the inspector. Before:

<img width="301" alt="Screenshot 2021-05-04 at 11 44 03" src="https://user-images.githubusercontent.com/1204802/116987065-e16a1b80-acce-11eb-9976-cc8f78794821.png">

After:

<img width="338" alt="Screenshot 2021-05-04 at 11 49 28" src="https://user-images.githubusercontent.com/1204802/116987071-e333df00-acce-11eb-9d22-7ba0a2c1feb1.png">

In the updated description text, can we link "your homepage" to `/wp-admin/options-reading.php`?

In some [recent mockups](https://user-images.githubusercontent.com/1204802/116987682-a6b4b300-accf-11eb-9adc-207dfaa4e950.png), I've gotten rid of some of the primary inspector panel entirely — you're unlikely to ever collapse it anyway, so might as well not have it. Can we do that here as well? Essentially move towards this instead:

<img width="368" alt="Screenshot 2021-05-04 at 11 54 23" src="https://user-images.githubusercontent.com/1204802/116987796-ca77f900-accf-11eb-9921-94849d57aae6.png">

This is something I'd like to see for all blocks, and it ties into #27331. I'd be happy to take a stab at that, but since this would be one of the first to do it, I'd be happy to explore that separately so as to not bog down this otherwise fine PR. What do you think?

Otherwise, this one is close to ready! Nice work 👌

gwwar committed:
Move home link translation strings to block.json add link to reading options

gwwar committed:
remove period

gwwar commented:
@jasmussen did we want the link to open in a new tab, or redirect in the editor?

<img width="290" alt="Screen Shot 2021-05-04 at 11 34 30 AM" src="https://user-images.githubusercontent.com/1270189/117052579-c39bb700-accc-11eb-86ac-82abcb25f30b.png">

🤔 It also looks like we have an [external link component](https://wordpress.github.io/gutenberg/?path=/story/components-externallink--default ) but not a link component. Anyone know why? It looks like it might be nice to have to make a11y cases a little easier to handle.

gwwar committed:
add home link settings inline instead of in panel body

gwwar committed:
do not add custom class twice

gwwar commented:
> In some recent mockups, I've gotten rid of some of the primary inspector panel entirely — you're unlikely to ever collapse it anyway, so might as well not have it. Can we do that here as well?

@jasmussen I gave it a try in https://github.com/WordPress/gutenberg/pull/30926/commits/a351c4b59d5c43335daaa366611cc4fe8a899b59

I think we'll want a component for this. Maybe something specific to the block-editor package since the block-card component exists there currently. Should we try it out in this PR or a follow up?

jasmussen commented:
Thanks so much for this:

<img width="397" alt="Screenshot 2021-05-05 at 09 53 08" src="https://user-images.githubusercontent.com/1204802/117111337-b8f22800-ad87-11eb-8c0f-1f56d6ab381d.png">

It's a fair bit cleaner, more obvious, less noisy, and balanced. 

<img width="243" alt="Screenshot 2021-05-05 at 09 53 48" src="https://user-images.githubusercontent.com/1204802/117111388-ceffe880-ad87-11eb-93cc-017d08023ba4.png">

What are the requirements for the "external link" icon? I had this instinct that it's primarily for links that open in a new tab, but I sense it's for anything that takes you away from the current page. In which case I need to find/create a link icon that is less obtrusive.

However in order to land this PR:

- We should add a period at the end of the description (Link to your homepage.)
- If you feel the new components for the non-collapsed inspector is best handled in a separate PR, we can do that (or just track it). Up to what you feel is best, I'd approve this in a panel called "Settings" if that is the better approach for now.

I'll put that external link icon on my todo.

Thanks Kerry!

mtias commented on packages/block-library/src/home-link/index.js:
```diff
@@ -0,0 +1,37 @@
+
+	description: createInterpolateElement(
+		_x( 'Link to <a>your homepage</a>', 'block description' ),
+		{ a: <ExternalLink href="options-reading.php" /> }
```
I don't think we should use an external link if we are linking within wp-admin. It's also probably worth noting that a static link like this would break in some contexts (clients like WP.com's Calypso, the mobile apps, etc).

gwwar replied:
Good point, @mtias. Is there a recommended pattern for allowing folks to override this value?

gwwar replied:
@Addison-Stavlo noted that we're using a slot/fill pattern in other areas https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/header/main-dashboard-button/index.js https://github.com/WordPress/gutenberg/blob/734e3871c1a9fc1085670a0fc5284baabfdf6b7b/packages/edit-post/src/components/header/index.js#L57-L59

I can maybe try this out too, though I suspect this may need to be generalized for any links in the block settings area.

mtias replied:
Yes, it can be tricky. It's one where a simpler filter seems better. In any case, I think it's fine not to add the url in this first pass and revisit more generally. We have a few other blocks that want to link to Settings > General or other admin places.

mtias commented on packages/block-library/src/home-link/block.json:
```diff
@@ -0,0 +1,39 @@
+		"rel": {
+			"type": "string"
+		},
+		"title": {
```
Do we need this attribute? Could we make it be the title of the site by default and not editable?

gwwar replied:
~~I asked for some quick feedback in https://wordpress.slack.com/archives/C02QB2JS7/p1620233144312500 and folks noted that it'd usually be "Home" instead of the site title.~~ Sorry misread feedback here :D

gwwar replied:
Hmm from https://make.wordpress.org/accessibility/handbook/content/link-destinations/#avoid-the-title-attribute-on-links maybe we can leave this unset for now?

mtias commented on packages/block-library/src/home-link/block.json:
```diff
@@ -0,0 +1,39 @@
+			"type": "boolean",
+			"default": false
+		},
+		"rel": {
```
What would be some use cases for `rel` on a home link?

gwwar replied:
🤔 browsing https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types it doesn't seem like there is anything relevant. I can remove this for v1, and we can see if folks ask to add it.

gwwar replied:
So I looked a bit and site title has `rel="home"` but there isn't any meaning to this in HTML5 spec. Is there a WP specific case for this?
 https://github.com/WordPress/gutenberg/blob/0ea3bcc05e33cfdc5a5945a0e408464357b73d87/packages/block-library/src/site-title/index.php#L23

mtias commented on packages/block-library/src/home-link/block.json:
```diff
@@ -0,0 +1,39 @@
+	"category": "design",
+	"parent": [ "core/navigation" ],
+	"title": "Home Link",
+	"keywords": [ "home", "menu", "navigation" ],
```
These keywords don't seem to add much. "home" should already be covered by the name. I'd probably remove them.

mtias commented:
@jasmussen not for this PR, but it'd be interesting to consider a setting to turn the link into a home icon.

gwwar commented:
> What are the requirements for the "external link" icon?

Oh no requirements @jasmussen I was asking if we wanted to open in a new tab or not. In terms of experience, folks might have unsaved changes, and a redirect would pop a warning.

I think it's fine if we prefer redirecting in the same tab. It is a bit odd that we don't have a plain link component to help with a11y requirements though.

jasmussen commented:
> Oh no requirements @jasmussen I was asking if we wanted to open in a new tab or not. In terms of experience, folks might have unsaved changes, and a redirect would pop a warning.

Oh, right, I was thinking of this link as opening in the same window:

<img width="303" alt="Screenshot 2021-05-05 at 19 12 44" src="https://user-images.githubusercontent.com/1204802/117181840-efef2a80-add5-11eb-9651-9901dc4cfc3c.png">

But that actually does open in a new window.

Yes, I'd think a plain link that opens in the same tab is fine, especially since you'd get a `confirm` warning if you had unsaved changes. My rule of thumb is to never open in a new tab unless there's a really good reason, and "preview in new window" is one such as you might want to compare. 

> It is a bit odd that we don't have a plain link component to help with a11y requirements though.

Just for my own education's sake, what would such requirements be? For opening in a new tab I understand the need for a textual equivalent to the icon, something that alerts it will open a new tab. But for vanilla links, instinctually I'd think screen readers had those covered.

gwwar commented:
> Just for my own education's sake, what would such requirements be?

Oh checking this again, maybe the requirement is only for external links? https://make.wordpress.org/accessibility/handbook/content/link-destinations/#the-link-opens-in-a-new-window-or-tab

I think we should be fine here 👍

gwwar commented:
With only one setting, it may make sense to remove the settings heading.

<img width="286" alt="Screen Shot 2021-05-05 at 10 58 11 AM" src="https://user-images.githubusercontent.com/1270189/117187588-edb6ad00-ad90-11eb-9f50-f4dd30a47da8.png">

gwwar committed:
remove title and rel fields

mtias commented:
I wouldn't be opposed to removing all settings entirely since it would be pretty rare to want the homepage to open in a new window / tab, and generally bad usability.

gwwar commented:
> I wouldn't be opposed to removing all settings entirely since it would be pretty rare to want the homepage to open in a new window / tab, and generally bad usability.

Fair enough, let's do it.

mtias commented:
We might want to tweak the description a bit to be less of a repetition of the title itself. Maybe give a hint to how it's not necessary if the site title is also a link.

> Create a link that always points to the homepage of the site. Usually not necessary if there is already a site title link present in the header.

gwwar committed:
remove open in new tab, back out link to settings

gwwar commented:
To recap recent changes:

* Removed rel, title, and opensInNewTab as block attributes
* I left title unset in the published view, due to https://make.wordpress.org/accessibility/handbook/content/link-destinations/#avoid-the-title-attribute-on-links
* `rel` is also unset, but Site Title has this set to `rel="home"` is there a specific WP use case for this? This doesn't have meaning in HTML5 spec.
* I removed the link to Reader Settings. We can follow up in a future PR. What makes this tricky is that we should allow folks to override this link, which might mean adding a filter, or a new slot/fill. See https://github.com/WordPress/gutenberg/pull/30926#discussion_r626765771

jasmussen reviewed (approved):
This is looking like a great start to me. Thanks for all the explorations, even if we ended up on something relatively minimalist:

<img width="1270" alt="Screenshot 2021-05-06 at 10 05 14" src="https://user-images.githubusercontent.com/1204802/117263505-9ed14c00-ae52-11eb-8b17-f6fb06e37733.png">

I'd encourage a code double-check, but since there's already been a few, I'll also let your confidence level press the green button when the time comes.

swissspidy commented on packages/block-library/src/home-link/index.php:
```diff
@@ -0,0 +1,178 @@
+ *
+ * @return string Returns the post content with the home url added.
+ */
+function render_block_core_home( $attributes, $content, $block ) {
```
```suggestion
function render_block_core_home_link( $attributes, $content, $block ) {
```

gwwar replied:
Thanks @swissspidy, I missed these in the home -> home_link rename. I've updated these in 10b9f65ab13fafad3794f30baefa5e147d6c3a9a

swissspidy commented on packages/block-library/src/home-link/index.php:
```diff
@@ -0,0 +1,178 @@
+ * @uses render_block_core_home()
+ * @throws WP_Error An WP_Error exception parsing the block definition.
+ */
+function register_block_core_home() {
```
```suggestion
function register_block_core_home_link() {
```

swissspidy commented on packages/block-library/src/home-link/index.php:
```diff
@@ -0,0 +1,178 @@
+	register_block_type_from_metadata(
+		__DIR__ . '/home-link',
+		array(
+			'render_callback' => 'render_block_core_home',
```
```suggestion
			'render_callback' => 'render_block_core_home_link',
```

swissspidy commented on packages/block-library/src/home-link/index.php:
```diff
@@ -0,0 +1,178 @@
+		)
+	);
+}
+add_action( 'init', 'register_block_core_home' );
```
```suggestion
add_action( 'init', 'register_block_core_home_link' );
```

swissspidy commented on packages/block-library/src/home-link/index.php:
```diff
@@ -0,0 +1,178 @@
+	$style_attribute = ( $colors['inline_styles'] . $font_sizes['inline_styles'] );
+	$css_classes     = trim( implode( ' ', $classes ) );
+
+	$wrapper_attributes = get_block_wrapper_attributes(
+		array(
+			'class' => $css_classes,
+			'style' => $style_attribute,
+		)
+	);
+
+	return $wrapper_attributes;
```
Attributes need to be escaped with `esc_attr()` ⚠️

gwwar committed:
update function names from _home to home_link

jasmussen commented:
> not for this PR, but it'd be interesting to consider a setting to turn the link into a home icon.

Did a quick exploration. My first instinct was to have an option directly in the home blocks block toolbar. But then I realized that the toggling on/off of bold/italic options might feel weird, and that we might want 3 options making it less of a toggle and more like a segmented control.

Here's a toggle:

<img width="1440" alt="Home icon" src="https://user-images.githubusercontent.com/1204802/117638500-2d133e00-b183-11eb-8e07-d3ac9d2e8fd6.png">

<img width="1440" alt="Home icon-1" src="https://user-images.githubusercontent.com/1204802/117638504-2e446b00-b183-11eb-9452-4de4aec455ef.png">

Here's a version with a segmented control:

<img width="1440" alt="Home icon alt" src="https://user-images.githubusercontent.com/1204802/117638532-36040f80-b183-11eb-9650-c04101021ca8.png">

Any preferences or feedback? Then I'll open a ticket.

gwwar commented:
> not for this PR, but it'd be interesting to consider a setting to turn the link into a home icon. [...] Did a quick exploration.

@jasmussen The segmented control looks great. If folks choose an icon only, we'd maybe need to allow folks to edit alt-text.

I think I addressed noted feedback. It looks like to move this very simplified version forward, we'll want a +1 code review. Does that sound about right?

jasmussen commented:
> I think I addressed noted feedback. It looks like to move this very simplified version forward, we'll want a +1 code review. Does that sound about right?

Would be a bonus, yes. But I would also apply the law of diminishing returns. I.e. a bit of silence means ship it.

mkaz reviewed (approved):
👍 This looks good and tests well for me. Nice first iteration, I think adding the icon and other options can be good next steps. :shipit:

gwwar commented:
Thanks y'all for the reviews ✨
